Docs: add ADR about `disabled` attributes (#1316)
CleverCloud/clever-components: docs/adr/adr-0029-issues-with-the-disabled-attribute.md

One thing to remember is that when set to `disabled`, native elements cannot receive focus.

## Different `disabled` implementations in our codebase

For our `cc-button` component, since there are many cases where pressing `Enter` or `Space` would trigger the API call, we did not want the focus to be lost when it was set to `disabled` during API calls.

This is why within the `cc-button` component, we rely on `aria-disabled` instead of `disabled` on the native `<button>`:

```html
<cc-button disabled>
    #shadow-dom
    <button aria-disabled="true">...</button>
</cc-button>
```

Since `aria-disabled` only handles the semantics, we prevent actions from going through with this:

```js
// in `cc-button.js` we disable the click handler if the component is disabled
_onClick(e) {
    e.stopPropagation();

    // we need to check that because we use aria-disabled which doesn't prevent the onclick event to be fired.
    if (this.disabled || this.skeleton || this.waiting) {
      return;
    }
  ...
}
```

On the contrary, within the `cc-input-*` components, we rely on the native `disabled` attribute:

> ```js

Reviewer: ```suggestion
```html
```

<cc-input-text disabled>
    #shadow-dom
    <input disabled>
</cc-input-text>
```

## Focus losses during API calls

### With `cc-input-*` components

During API calls, we always set `cc-input-*` components in `disabled`.
This may lead to focus being lost in the following scenario:

1. The user focuses the `input` within `cc-input-text`,
2. The user enters the value of their choice,
3. The user presses `Enter` to submit the form,
4. The native `input` being focused is set to `disabled`,

> 5. The focus is lost (usually reset to the body but this is browser heuristics territory).

Reviewer 2: **🤓 nitpick:** to mimic L82

```suggestion
5. The focus is lost (usually reset to the `body` but this is browser heuristics territory).
```

   - For `cc-button` components:
    - Use a new `waiting` attribute/property instead of `disabled`,
    - Throw errors if both `disabled` and `waiting` are set to `true` at the same time.

2. Only use `disabled` when:
   - The control should be permanently disabled
   - The control should not participate in form validation
   - The control should be completely non-interactive

## Consequences

### Positive

- Better focus management
- Improved accessibility
- More predictable behavior across browsers
- Clearer separation between temporary and permanent disabled states

### Negative

- Need to maintain multiple ways to handle disabled-like states
- Slightly more complex implementation
- Requires clear documentation and guidelines for developers

## Implementation Guidelines

1. Form Controls:
```javascript
// During API calls
<cc-input-text readonly>
// For permanent disability
<cc-input-text disabled>
```

2. Buttons:
```javascript
// During API calls
<cc-button waiting>
// For permanent disability
<cc-button disabled>
```

3. Error Prevention:
```javascript
// We prevent using both states simultaneously

> <cc-button waiting disabled> // Will throw console.error

Reviewer: suggestion:
```suggestion
<cc-button waiting disabled> // Will not render and an error will be printed to console
```
